File: backend/app/utils/bitset.py
from __future__ import annotations

from typing import List, Tuple

from app.core.constants import (
    BITS_PER_TAG as BITS_PER_TAG,
    BYTES_PER_DAY as BYTES_PER_DAY,
    MINUTES_PER_SLOT as MINUTES_PER_SLOT,
    SLOTS_PER_DAY as SLOTS_PER_DAY,
    TAG_BYTES_PER_DAY as TAG_BYTES_PER_DAY,
    TAG_NO_TRAVEL as TAG_NO_TRAVEL,
    TAG_NONE as TAG_NONE,
    TAG_ONLINE_ONLY as TAG_ONLINE_ONLY,
    TAG_RESERVED as TAG_RESERVED,
)


def new_empty_bits() -> bytes:
    return bytes(BYTES_PER_DAY)


def new_empty_tags() -> bytes:
    return bytes(TAG_BYTES_PER_DAY)


def pack_indexes(indexes: List[int]) -> bytes:
    b = bytearray(BYTES_PER_DAY)
    for idx in indexes:
        if not (0 <= idx < SLOTS_PER_DAY):
            raise ValueError(f"index out of range: {idx}")
        byte_i = idx // 8
        bit_i = idx % 8
        b[byte_i] |= 1 << bit_i
    return bytes(b)


def unpack_indexes(bits: bytes) -> List[int]:
    if len(bits) != BYTES_PER_DAY:
        raise ValueError(f"bits length must be {BYTES_PER_DAY}")
    out: List[int] = []
    for byte_i, val in enumerate(bits):
        for bit_i in range(8):
            idx = byte_i * 8 + bit_i
            if idx >= SLOTS_PER_DAY:
                break
            if (val >> bit_i) & 1:
                out.append(idx)
    return out


def toggle_index(bits: bytes, idx: int, value: bool) -> bytes:
    if len(bits) != BYTES_PER_DAY:
        raise ValueError(f"bits length must be {BYTES_PER_DAY}")
    if not (0 <= idx < SLOTS_PER_DAY):
        raise ValueError("index out of range")
    b = bytearray(bits)
    byte_i = idx // 8
    bit_i = idx % 8
    if value:
        b[byte_i] |= 1 << bit_i
    else:
        b[byte_i] &= ~(1 << bit_i)
    return bytes(b)


def get_slot_tag(tags: bytes, slot: int) -> int:
    if len(tags) != TAG_BYTES_PER_DAY:
        raise ValueError(f"tags length must be {TAG_BYTES_PER_DAY}")
    if not (0 <= slot < SLOTS_PER_DAY):
        raise ValueError("slot out of range")
    bit_offset = slot * BITS_PER_TAG
    byte_idx = bit_offset // 8
    bit_pos = bit_offset % 8
    return (tags[byte_idx] >> bit_pos) & 0b11


def set_slot_tag(tags: bytes, slot: int, tag: int) -> bytes:
    if len(tags) != TAG_BYTES_PER_DAY:
        raise ValueError(f"tags length must be {TAG_BYTES_PER_DAY}")
    if not (0 <= slot < SLOTS_PER_DAY):
        raise ValueError("slot out of range")
    if not (TAG_NONE <= tag <= TAG_RESERVED):
        raise ValueError("tag must be 0-3")
    bit_offset = slot * BITS_PER_TAG
    byte_idx = bit_offset // 8
    bit_pos = bit_offset % 8
    b = bytearray(tags)
    b[byte_idx] &= ~(0b11 << bit_pos)
    b[byte_idx] |= (tag & 0b11) << bit_pos
    return bytes(b)


def set_range_tag(tags: bytes, start_slot: int, count: int, tag: int) -> bytes:
    if len(tags) != TAG_BYTES_PER_DAY:
        raise ValueError(f"tags length must be {TAG_BYTES_PER_DAY}")
    if count <= 0:
        raise ValueError("count must be greater than 0")
    if start_slot < 0 or start_slot + count > SLOTS_PER_DAY:
        raise ValueError("range out of bounds")
    if not (TAG_NONE <= tag <= TAG_RESERVED):
        raise ValueError("tag must be 0-3")
    result = bytearray(tags)
    for i in range(count):
        slot = start_slot + i
        bit_offset = slot * BITS_PER_TAG
        byte_idx = bit_offset // 8
        bit_pos = bit_offset % 8
        result[byte_idx] &= ~(0b11 << bit_pos)
        result[byte_idx] |= (tag & 0b11) << bit_pos
    return bytes(result)


def get_range_tag(tags: bytes, start_slot: int, count: int) -> int | None:
    if count <= 0:
        raise ValueError("count must be greater than 0")
    first = get_slot_tag(tags, start_slot)
    for i in range(1, count):
        if get_slot_tag(tags, start_slot + i) != first:
            return None
    return first


def is_tag_compatible(tag: int, location_type: str) -> bool:
    if tag == TAG_NONE:
        return True
    if tag == TAG_ONLINE_ONLY:
        return location_type == "online"
    if tag == TAG_NO_TRAVEL:
        return location_type in {"online", "instructor_location"}
    return False  # TAG_RESERVED (3) intentionally blocks all formats


def windows_from_bits(bits: bytes) -> List[Tuple[str, str]]:
    """Return merged windows as ('HH:MM:SS','HH:MM:SS') tuples."""
    idxs = unpack_indexes(bits)
    if not idxs:
        return []
    windows: List[Tuple[int, int]] = []
    start = prev = idxs[0]
    for idx in idxs[1:]:
        if idx == prev + 1:
            prev = idx
            continue
        windows.append((start, prev + 1))
        start = prev = idx
    windows.append((start, prev + 1))

    def idx_to_time(i: int) -> str:
        minutes = i * MINUTES_PER_SLOT
        hh = minutes // 60
        mm = minutes % 60
        return f"{hh:02d}:{mm:02d}:00"

    return [(idx_to_time(s), idx_to_time(e)) for s, e in windows]


def bits_from_windows(windows: List[Tuple[str, str]]) -> bytes:
    """Inverse of windows_from_bits for 'HH:MM[:SS]' strings."""

    def time_to_index(t: str, *, is_end: bool = False) -> int:
        parts = t.split(":")
        if len(parts) < 2:
            raise ValueError(f"invalid time format: {t}")
        hh, mm = int(parts[0]), int(parts[1])
        if is_end and hh == 0 and mm == 0:
            return SLOTS_PER_DAY
        return (hh * 60 + mm) // MINUTES_PER_SLOT

    idxs: List[int] = []
    for start, end in windows:
        s = time_to_index(start)
        e = time_to_index(end, is_end=True)
        if not (0 <= s <= e <= SLOTS_PER_DAY):
            raise ValueError(f"window out of bounds: {start}-{end}")
        idxs.extend(range(s, e))
    return pack_indexes(idxs)

File: frontend/lib/calendar/bitset.ts
export const SLOTS_PER_DAY = 288;
export const MINUTES_PER_SLOT = 5;
export const BYTES_PER_DAY = 36;
export const BOOKING_START_STEP_MINUTES = 15;
export const AVAILABILITY_CELL_MINUTES = 30;
export const BITS_PER_CELL = 6; // AVAILABILITY_CELL_MINUTES / MINUTES_PER_SLOT
export const BITS_PER_TAG = 2;
export const TAG_BYTES_PER_DAY = Math.ceil((SLOTS_PER_DAY * BITS_PER_TAG) / 8);
export const TAG_NONE = 0;
export const TAG_ONLINE_ONLY = 1;
export const TAG_NO_TRAVEL = 2;
export const TAG_RESERVED = 3;
export type DayBits = Uint8Array;
export type DayTags = Uint8Array;
export type FormatTag =
  | typeof TAG_NONE
  | typeof TAG_ONLINE_ONLY
  | typeof TAG_NO_TRAVEL
  | typeof TAG_RESERVED;

export function newEmptyBits(): DayBits {
  return new Uint8Array(BYTES_PER_DAY);
}

export function newEmptyTags(): DayTags {
  return new Uint8Array(TAG_BYTES_PER_DAY);
}

export function idx(hh: number, mm: number): number {
  return Math.floor((hh * 60 + mm) / MINUTES_PER_SLOT);
}

export function toWindows(
  bits: DayBits,
): { start_time: string; end_time: string }[] {
  const on: number[] = [];
  for (let b = 0; b < BYTES_PER_DAY; b += 1) {
    const byteValue = bits[b] ?? 0;
    for (let bit = 0; bit < 8; bit += 1) {
      const i = b * 8 + bit;
      if (i >= SLOTS_PER_DAY) break;
      if ((byteValue >> bit) & 1) on.push(i);
    }
  }
  if (on.length === 0) return [];

  const toTime = (slotIdx: number) => {
    const totalMinutes = slotIdx * MINUTES_PER_SLOT;
    const H = Math.floor(totalMinutes / 60);
    const M = totalMinutes % 60;
    return `${String(H).padStart(2, "0")}:${String(M).padStart(
      2,
      "0",
    )}:00`;
  };

  const out: { start_time: string; end_time: string }[] = [];
  const firstSlot = on[0]!;
  let start = firstSlot;
  let prev = firstSlot;
  for (let k = 1; k < on.length; k += 1) {
    const current = on[k]!;
    if (current === prev + 1) {
      prev = current;
      continue;
    }
    out.push({ start_time: toTime(start), end_time: toTime(prev + 1) });
    start = current;
    prev = current;
  }
  out.push({ start_time: toTime(start), end_time: toTime(prev + 1) });
  return out;
}

export function fromWindows(
  windows: { start_time: string; end_time: string }[],
): DayBits {
  const bits = newEmptyBits();
  const toIdx = (timeStr: string, isEndTime: boolean = false): number => {
    const [H, M] = timeStr.split(":").map(Number);
    // Midnight (00:00) as end time means end-of-day (slot 288)
    if (isEndTime && H === 0 && M === 0) {
      return SLOTS_PER_DAY;
    }
    return idx(Number(H), Number(M));
  };

  for (const { start_time, end_time } of windows) {
    const start = Math.max(0, Math.min(SLOTS_PER_DAY, toIdx(start_time, false)));
    const end = Math.max(0, Math.min(SLOTS_PER_DAY, toIdx(end_time, true)));
    for (let i = start; i < end; i += 1) {
      const byte = Math.floor(i / 8);
      if (byte >= bits.length) break;
      const bit = i % 8;
      const current: number = bits[byte] ?? 0;
      bits[byte] = current | (1 << bit);
    }
  }
  return bits;
}

export function toggle(bits: DayBits, slotIndex: number, on: boolean): DayBits {
  const next = bits.slice();
  const byte = Math.floor(slotIndex / 8);
  if (byte >= next.length || slotIndex < 0 || slotIndex >= SLOTS_PER_DAY) {
    return next;
  }
  const bit = slotIndex % 8;
  const current: number = next[byte] ?? 0;
  next[byte] = on ? current | (1 << bit) : current & ~(1 << bit);
  return next;
}

export function getSlotTag(tags: DayTags, slot: number): FormatTag {
  if (tags.length !== TAG_BYTES_PER_DAY) {
    throw new Error(`tags length must be ${TAG_BYTES_PER_DAY}`);
  }
  if (slot < 0 || slot >= SLOTS_PER_DAY) {
    throw new Error('slot out of range');
  }
  const bitOffset = slot * BITS_PER_TAG;
  const byteIdx = Math.floor(bitOffset / 8);
  const bitPos = bitOffset % 8;
  return (((tags[byteIdx] ?? 0) >> bitPos) & 0b11) as FormatTag;
}

export function setSlotTag(tags: DayTags, slot: number, tag: FormatTag): DayTags {
  if (tags.length !== TAG_BYTES_PER_DAY) {
    throw new Error(`tags length must be ${TAG_BYTES_PER_DAY}`);
  }
  if (slot < 0 || slot >= SLOTS_PER_DAY) {
    throw new Error('slot out of range');
  }
  if (tag < TAG_NONE || tag > TAG_RESERVED) {
    throw new Error('tag must be 0-3');
  }
  const next = tags.slice();
  const bitOffset = slot * BITS_PER_TAG;
  const byteIdx = Math.floor(bitOffset / 8);
  const bitPos = bitOffset % 8;
  next[byteIdx] = (next[byteIdx] ?? 0) & ~(0b11 << bitPos);
  next[byteIdx] = (next[byteIdx] ?? 0) | ((tag & 0b11) << bitPos);
  return next;
}

export function setRangeTag(tags: DayTags, startSlot: number, count: number, tag: FormatTag): DayTags {
  if (tags.length !== TAG_BYTES_PER_DAY) {
    throw new Error(`tags length must be ${TAG_BYTES_PER_DAY}`);
  }
  if (count <= 0) {
    throw new Error('count must be greater than 0');
  }
  if (startSlot < 0 || startSlot + count > SLOTS_PER_DAY) {
    throw new Error('range out of bounds');
  }
  if (tag < TAG_NONE || tag > TAG_RESERVED) {
    throw new Error('tag must be 0-3');
  }
  const next = tags.slice();
  for (let i = 0; i < count; i += 1) {
    const slot = startSlot + i;
    const bitOffset = slot * BITS_PER_TAG;
    const byteIdx = Math.floor(bitOffset / 8);
    const bitPos = bitOffset % 8;
    next[byteIdx] = (next[byteIdx] ?? 0) & ~(0b11 << bitPos);
    next[byteIdx] = (next[byteIdx] ?? 0) | ((tag & 0b11) << bitPos);
  }
  return next;
}

export function getRangeTag(tags: DayTags, startSlot: number, count: number): FormatTag | null {
  if (count <= 0) {
    throw new Error('count must be greater than 0');
  }
  const first = getSlotTag(tags, startSlot);
  for (let i = 1; i < count; i += 1) {
    if (getSlotTag(tags, startSlot + i) !== first) {
      return null;
    }
  }
  return first;
}

/** Toggle a range of consecutive slots (used by 30-min grid cells -> 6 bits). */
export function toggleRange(bits: DayBits, startSlot: number, count: number, on: boolean): DayBits {
  const next = bits.slice();
  for (let i = 0; i < count; i += 1) {
    const slot = startSlot + i;
    const byte = Math.floor(slot / 8);
    if (byte >= next.length || slot < 0 || slot >= SLOTS_PER_DAY) continue;
    const bit = slot % 8;
    const current: number = next[byte] ?? 0;
    next[byte] = on ? current | (1 << bit) : current & ~(1 << bit);
  }
  return next;
}

/** Check if ALL bits in a range are set (for 30-min cell display). */
export function isRangeSet(bits: DayBits | undefined, startSlot: number, count: number): boolean {
  if (!bits) return false;
  for (let i = 0; i < count; i += 1) {
    const slotIndex = startSlot + i;
    const byte = Math.floor(slotIndex / 8);
    const bit = slotIndex % 8;
    if ((((bits[byte] ?? 0) >> bit) & 1) === 0) return false;
  }
  return true;
}
